File: backend/src/main/java/marcotumminia/capstone/codifyRef/annuncio/AnnuncioService.java
package marcotumminia.capstone.codifyRef.annuncio;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import marcotumminia.capstone.codifyRef.exceptions.NotFoundException;
import marcotumminia.capstone.codifyRef.utente.Utente;
import marcotumminia.capstone.codifyRef.utente.UtenteRepository;
import marcotumminia.capstone.codifyRef.utente.UtenteService;

@Service
public class AnnuncioService {

    @Autowired
    private AnnuncioRepository annuncioRepository;
    
    @Autowired
    private UtenteService utenteService;
    
    @Autowired
    private UtenteRepository utenteRepository;
    
    
    ////PER CREARE L'ANNUNCIO
    public Annuncio createAnnuncio(AnnuncioPayload payload) {
    	
        Utente utenteAutenticato = utenteService.getCurrentUser();

        Annuncio annuncio = new Annuncio(
            payload.getTitolo(),
            payload.getDescrizione(),
            payload.getCategoria(),
            StatoAnnuncio.APERTO,
            payload.getBudgetPrevisto(),
            LocalDate.now()  
            
        );
        annuncio.setUtente(utenteAutenticato);

        return save(annuncio);
    }

    public Annuncio save(Annuncio annuncio) {
        return annuncioRepository.save(annuncio);
    }
    
    
    
    /////PER TROVARE UN ANNUNCIO BY ID    
    public Annuncio findAnnuncioById(UUID id) {
        Optional<Annuncio> optionalAnnuncio = annuncioRepository.findAnnuncioById(id);
        
        if (optionalAnnuncio.isPresent()) {
            return optionalAnnuncio.get();
        } else {
      
            return null;
        }
    }
    
    
    
    /////PAGINAZIONE DEGLI ANNUNCI PER HOME ANNUNCI
    public Page<Annuncio> find(int page, int size, String sort){
    	Pageable pageable = PageRequest.of(page, size, Sort.by(sort));
    	return annuncioRepository.findAll(pageable);
    }

   
    
    ///// METODO PER RECUPERARE TUTTI GLI ANNUNCI DI UN UTENTE
    public List<Annuncio> findAnnunciByUtente(UUID utenteId) {
        Utente utente = utenteRepository.findById(utenteId)
                .orElseThrow(() -> new NotFoundException("Utente non trovato con l'ID: " + utenteId));
        
        return annuncioRepository.findByUtente(utente);
    }
    
	/////////METODO ELIMINARE UN ANNUNCIO SPECIFICO
    public void findByIdAndDelete(UUID id) {
    	if (annuncioRepository.existsById(id)) {
    		annuncioRepository.deleteById(id);
    	} else {  throw new NotFoundException("Annuncio non trovato con l'ID: " + id);
    	}
    }
    
    
    /////////METODO PER FILTRARE LA BACHECA DI ANNUNCIO PER CATEGORIA
    public Page<Annuncio> findByCategoria(CategoriaAnnuncio categoria, int page, int size, String sort) {
        Pageable pageable = PageRequest.of(page, size, Sort.by(sort));
        return annuncioRepository.findByCategoria(categoria, pageable);
    }

        
}

File: frontend/src/app/service/annuncio.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Annuncio } from '../components/annunci/annuncio';
import * as UUID from 'uuid';

@Injectable({
  providedIn: 'root',
})
export class AnnuncioService {
  private apiUrl = 'http://localhost:3001/annunci';

  constructor(private http: HttpClient) {}

  //METODO PER CREARE L'ANNUNCIO
  createAnnuncio(payload: any): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/nuovo-annuncio`, payload);
  }

  //METODO PER RECUPERARE TUTTI GLI ANNUNCI (BACHECA PAGINATA)
  getAllAnnunci(
    page: number,
    size: number,
    sort: string
  ): Observable<Annuncio[]> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sort', sort);

    return this.http.get<Annuncio[]>(`${this.apiUrl}/bacheca-annunci`, {
      params,
    });
  }

  //METODO PER FILTRARE GLI ANNUNCI IN BACHECA PER CATEGORIA
  getAnnunciByCategoria(categorie: string[]) {
    const categorieString = categorie.join(',');
    return this.http.get<Annuncio[]>(
      `${this.apiUrl}/annunci-per-categoria?categoria=${categorieString}`
    );
  }

  //METODO PER RECUPERARE I DETTAGLI DI UN ANNUNCIO SPECIFICO
  getAnnuncioById(annuncioId: string): Observable<Annuncio> {
    return this.http.get<Annuncio>(`${this.apiUrl}/${annuncioId}`);
  }

  // Metodo per recuperare gli annunci di un utente
  getAnnunciByUtente(utenteId: string): Observable<Annuncio[]> {
    return this.http.get<Annuncio[]>(`${this.apiUrl}/utente/${utenteId}`);
  }
}
